### ISSUE TITLE
Category tree move ends with dead lock

### ISSUE THREAD

**Issue by Detzler** [magento/magento2#12380]
### Preconditions

1. Magento2.2
2. PHP7.1

### Steps to reproduce

1. Execute code:
```
$this->collectionFactory->create();  
$category->setName('test');
$category->save();
$category->move(2, null);
```

### Expected result

1. Category moves in tree

### Actual result

1. Script execution aborts with error:

> SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: DELETE `url_rewrite`  FROM `url_rewrite` WHERE (`entity_id` IN ('2008')) AND (`entity_type` IN ('category'))

### Hints
The move method in `vendor/magento/module-catalog/Model/Category.php`
n dispatches an event via:
`$this->_eventManager->dispatch($this->_eventPrefix . '_move_after', $eventParams);`

`vendor/magento/module-catalog-url-rewrite/Observer/CategoryProcessUrlRewriteMovingObserver.php` observes the event

When commenting out the event dispatch, the move method works fine without deadlocks.

**Comment by magento-engcom-team** [comment]
@Detzler, thank you for your report.
We were not able to reproduce this issue by following the steps you provided. Please provide more detailed steps to reproduce or try to reproduce this issue on a clean installation or latest release.

**Comment by magento-engcom-team** [comment]
@Detzler, we are closing this issue due to inactivity. If you'd like to update it, please reopen the issue.

**Comment by okovalov-commer** [comment]
I have this error too

> SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: DELETE `url_rewrite`  FROM `url_rewrite` WHERE (`entity_id` IN ('190')) AND (`entity_type` IN ('product')) AND (`redirect_type` IN (0)) AND (`store_id` IN ('1'))"}

caused by 

```
        $existingCategoryIds = $product->getCategoryIds();
        
       $categoryIdsToSet = array_unique(array_merge($existingCategoryIds, $newCategoryIds));

        $this->categoryLinkManagement->assignProductToCategories(
            $product->getSku(),
            $categoryIdsToSet
        );
```

**Comment by niteshv0909** [comment]
Is this fixed ??  I am getting this in magento 2.2.4.

**Comment by jasonhildebrand** [comment]
I'm using magento 2.2.11 and getting deadlock in assignProductToCategories().

I have a console script to assign categories for 20000 products.  It fails approx every 1000 records.

In my case I get this exception chain:
                                                                 
```
[Magento\Framework\Exception\CouldNotSaveException]            
Could not save product "6379" with position 0 to category 807  
                                                                                                                                                                                               
[Magento\Framework\DB\Adapter\DeadlockException]                                                                            
SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction, query was  
  : INSERT  INTO `catalog_category_product` (`category_id`,`product_id`,`position`) VALUES (?, ?, ?)                          
                                                                                                                                                                                                                                                            
[Zend_Db_Statement_Exception]                                                                                               
SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction, query was  
  : INSERT  INTO `catalog_category_product` (`category_id`,`product_id`,`position`) VALUES (?, ?, ?)                          
                                                                                                                                                                                                                                                 
[PDOException]                                                                                                   
SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction  

```                                                                                                                   

I am using Percona 5.7.

Since this function is prone to deadlock, my workaround is to catch and re-attempt the call using this pattern:

```
        $attempts = 0;
        while ($attempts < 5) {
            try {
                $attempts += 1;
                $this->categoryLinkManagementInterface->assignProductToCategories($sku, $assignedCategories);
                break;
            } catch (\Magento\Framework\Exception\CouldNotSaveException $exception) {
                if (strpos(get_class($exception->getPrevious()), 'Deadlock') !== false) {
                    // Caught deadlock; try again
                    continue;
                }
                throw $exception;
            }
        }

```
